### PR TITLE
Masked Patch Prediction "Suggested in #63" Work in Progress

### PR DESCRIPTION
**Pull request by zankner**
Implementation of masked patch prediction for self-supervised learning.

### CODE REVIEW

File: vit_pytorch/masked_prediction_loss.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            for masked_patch in masked_patches_dims:
+                height_offset, width_offset = masked_patch
+                extracted_patch = targets[target, :,
+                                        height_offset:height_offset + self.patch_size,
+                                        width_offset:width_offset + self.patch_size]
+                target_patches.append(extracted_patch)
+
+        target_patches_tensor = torch.stack(target_patches)
+        target_patches_tensor = (target_patches_tensor // 0.34).long()
+        encoded_targets = F.one_hot(target_patches_tensor, 3)
```
Comment (lucidrains) [inline vit_pytorch/masked_prediction_loss.py:33]: Also a bit confused by this line

File: vit_pytorch/masked_prediction_loss.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        encoded_targets = torch.reshape(encoded_targets, [n, w, h, c * e])
+        mean_targets = torch.mean(encoded_targets.float(),
+                                dim=[1, 2],
+                                keepdim=True).view(n, c * e)
+        return mean_targets
+
+    def _transform_outputs(self, outputs, masked_patches):
+        output_dim = outputs.shape[-1]
+
+        masked_patches_shifted = [
```
Comment (lucidrains) [inline vit_pytorch/masked_prediction_loss.py:44]: is this being shifted because of the CLS token?

File: vit_pytorch/vit_pytorch.py [hunk 3]
```diff
@@ -109,15 +110,19 @@ def __init__(self, *, image_size, patch_size, num_classes, dim, depth, heads, ml
             nn.Linear(dim, num_classes)
         )
 
-    def forward(self, img, mask = None):
+    def forward(self, img, mask = None, prediction_mask = None):
```
Comment (lucidrains) [inline vit_pytorch/vit_pytorch.py:113]: i think, with a wrapper class, we can actually avoid modifying ViT altogether

File: vit_pytorch/masked_prediction_loss.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        target_patches = []
+        for target in range(targets.shape[0]):
+            for masked_patch in masked_patches_dims:
+                height_offset, width_offset = masked_patch
+                extracted_patch = targets[target, :,
+                                        height_offset:height_offset + self.patch_size,
+                                        width_offset:width_offset + self.patch_size]
+                target_patches.append(extracted_patch)
+
+        target_patches_tensor = torch.stack(target_patches)
```
Comment (lucidrains) [inline vit_pytorch/masked_prediction_loss.py:31]: all the above can probably be done with einops and then indexing

```python
from einops import rearrange

target_patches_tensor = rearrange(targets, 'b c (h p1) (w p2) -> b (h w) c h w', p1 = self.patch_size, p2 = self.patch_size)
target_patches_tensor = target_patches_tensor[:, masked_patches]
```

or something like that

**Comment by lucidrains** [comment]
@zankner Hi Zack! Thanks for the PR! Left you some comments - I think the PR may require some back and forth still, but I'd like to see it added. One question I have is, this leaves the generation of the random masks up to the user?

**Comment by lucidrains** [comment]
@zankner Hi Zack! Thanks for the PR! Left you some comments - I think the PR may require some back and forth. One question I have is, this leaves the generation of the random masks up to the user?

**Comment by lucidrains** [comment]
@zankner Is your code aimed at the implementation where the masked tokens predict the mean pixel value?

**Comment by zankner** [comment]
> @zankner Hi Zack! Thanks for the PR! Left you some comments - I think the PR may require some back and forth. One question I have is, this leaves the generation of the random masks up to the user?

As of right now it would but the PR still isn't finished yet. Just wanted to get the outline out to get your feedback on the direction.

**Comment by zankner** [comment]
> @zankner Is your code aimed at the implementation where the masked tokens predict the mean pixel value?

Yeah exactly.

File: vit_pytorch/masked_prediction_loss.py (hunk 1 above)
Comment (zankner) [reply vit_pytorch/masked_prediction_loss.py:33]: The one hot line?

File: vit_pytorch/masked_prediction_loss.py (hunk 2 above)
Comment (zankner) [reply vit_pytorch/masked_prediction_loss.py:44]: Yeah

File: vit_pytorch/masked_prediction_loss.py (hunk 4 above)
Comment (zankner) [reply vit_pytorch/masked_prediction_loss.py:31]: Ok cool, I haven't seen einops before but I'll look into that. Thanks for the suggestions!

**Comment by lucidrains** [comment]
@zankner very cool! I'll show you two of my self-supervised repos i was planning on repurposing for ViT

1. masked language modeling - https://github.com/lucidrains/mlm-pytorch
2. electra - https://github.com/lucidrains/electra-pytorch

i think the wrapper technique would work well with ViT, and the masked token would belong as a parameter under that wrapper

anyways, i'll let you think about it some more, let me know if you ever want me to step in and finish it

btw, i'm borned and raised around Boston too :) I'm from Andover :)

**Comment by zankner** [comment]
> @zankner very cool! I'll show you two of my self-supervised repos i was planning on repurposing for ViT
> 
> 1. masked language modeling - https://github.com/lucidrains/mlm-pytorch
> 2. electra - https://github.com/lucidrains/electra-pytorch
> 
> i think the wrapper technique would work well with ViT, and the masked token would belong as a parameter under that wrapper
> 
> anyways, i'll let you think about it some more, let me know if you ever want me to step in and finish it
> 
> btw, i'm born and raised around Boston too :) I'm from Andover :)

@lucidrains Awesome! I'll give those repos a check. Yeah, the task is basically the same as masked language modeling so I'm sure whatever is there will be perfect. 

That's awesome that your from Boston. It's a little bit cold right now, hopefully, you're in a place with some better weather 😂

**Comment by lucidrains** [comment]
@zankner ha, I'm in San Francisco now! got tired of the snow :) stay warm and drive safe!

File: vit_pytorch/masked_prediction_loss.py (hunk 1 above)
Comment (lucidrains) [reply vit_pytorch/masked_prediction_loss.py:33]: ohh i get it, this is for the 3 bit values for each channel? do you think for discretization, perhaps https://pytorch.org/docs/stable/generated/torch.searchsorted.html would be well-suited?

Comment (zankner) [reply vit_pytorch/masked_prediction_loss.py:33]: Yeah, search sorted would definitely work for that. Do you think that search sorted would perform better than the current method with one hot?

Comment (zankner) [reply vit_pytorch/masked_prediction_loss.py:33]: Just asking out of curiosity btw (trying to learn how to write better code)

Comment (lucidrains) [reply vit_pytorch/masked_prediction_loss.py:33]: yea i think so! one of the methods based on searchsorted, bucketize https://pytorch.org/docs/stable/generated/torch.bucketize.html is made just for this purpose

Comment (lucidrains) [reply vit_pytorch/masked_prediction_loss.py:33]: it would be a lot clearer just to use it

**Comment by zankner** [comment]
@lucidrains Hey, the last push should have a mostly working wrapper implementation. Still had to modify the forward call in vit module as the image has to be reshaped first in the forward pass of the wrapper. Reason being that the tokens are no longer just ids for embedding layer, but are instead patches of the image which affects the way that the mask token gets added to the input.

btw - do you have a preferred style guide for me to use with the code?

File: vit_pytorch/__init__.py [hunk 5]
```diff
@@ -1 +1,3 @@
 from vit_pytorch.vit_pytorch import ViT
+from vit_pytorch.mpp_loss import MPPLoss
```
Comment (lucidrains) [inline vit_pytorch/__init__.py:2]: You can just place the `MPPLoss` into the same file as `mpp_pytorch`, since it likely will not be reused

File: vit_pytorch/vit_pytorch.py [hunk 6]
```diff
@@ -109,10 +109,13 @@ def __init__(self, *, image_size, patch_size, num_classes, dim, depth, heads, ml
             nn.Linear(dim, num_classes)
         )
 
-    def forward(self, img, mask = None):
-        p = self.patch_size
+    def forward(self, img, mask = None, mpp = False):
```
Comment (lucidrains) [inline vit_pytorch/vit_pytorch.py:112]: i think you can actually avoid modifying ViT altogether if you just reach into the ViT class and invoke the `.transformer`

File: test.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import torch
+from vit_pytorch import MPP, ViT
+# from vit_pytorch import ViT, MaskedPredictionLoss
+
+v = ViT(
+    image_size = 256,
+    patch_size = 32,
+    num_classes = 3,
```
Comment (lucidrains) [inline test.py:8]: so my understanding was that for each of the channels (r, g, b), they found the average, and then discretize to 3 bit, which would yield 9 bits, and thus be predicting 512 classes here

but correct me if i'm wrong!

**Review by lucidrains — COMMENTED** [review]
looking better, but I left some comments!

this could also benefit from a section in the readme eventually!

File: test.py (hunk 7 above)
Comment (zankner) [reply test.py:8]: I'm not sure if this logic works, but I was thinking that I can just keep it as 3 channel, but have each channel be discretized to 0, 1, or 2 to replicate the 3 bits for each channel

Comment (zankner) [reply test.py:8]: Its probably just more straightforward and avoids any confusion to just discretize to 3 bits for each channel though

File: vit_pytorch/vit_pytorch.py (hunk 6 above)
Comment (zankner) [reply vit_pytorch/vit_pytorch.py:112]: I'm not sure if this is right, but doesn't the transformer take the embeddings of inputs after they have gone through a linear transformation and with the addition of the cls token. If I were to just invoke `.transformer` from the wrapper, I think the linear transformation and cls token would have to be parameters of the mpp wrapper module.

File: vit_pytorch/__init__.py (hunk 5 above)
Comment (zankner) [reply vit_pytorch/__init__.py:2]: Sounds good

**Comment by zankner** [comment]
> looking better, but I left some comments!
> 
> this could also benefit from a section in the readme eventually!

Yeah that makes a lot of sense about the readme

File: vit_pytorch/vit_pytorch.py (hunk 6 above)
Comment (lucidrains) [reply vit_pytorch/vit_pytorch.py:112]: ohh yea, you are right... hmm I think you should just reach in and grab the cls token and linear transform as well. the benefits of leaving the base class unchanged outweighs the extra complexity in the wrapper, imo

File: test.py (hunk 7 above)
Comment (lucidrains) [reply test.py:8]: yup, just have each channel discretize to 0, 1, 2, then add one, cumulative product, subtract one to get it into the range of 0 - 512

Comment (lucidrains) [reply test.py:8]: i don't think that works, i think you'll want to discretize to 8 values for each channel, then do a cumulative product of the three channels to get it into the range of 0 - 512

**Comment by zankner** [comment]
@lucidrains just pushed a change that fixes loss and leaves vit forward unchanged. Sorry to bother you with this question again, but is there any style guide you prefer like yapf or pep8?

**Comment by lucidrains** [comment]
@zankner looking better! i think remaining is really discretizing the RGB values to 512, and various cleanup and readme :)

hmm, I like my style the best :D perk of owning your own repos in a non-corporate environment ;)

**Comment by lucidrains** [comment]
@zankner just use whatever style you like (but leave the rest of the repository code unchanged)

File: vit_pytorch/mpp_pytorch.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    def forward(self, predicted_patches, target, mask):
+        # reshape target to patches
+        p = self.patch_size
+        target = rearrange(target, "b c (h p1) (w p2) -> b (h w) c (p1 p2) ", p1 = p, p2 = p)
+
+        avg_target = target.mean(dim=3)
+
+        channel_bins = torch.tensor([0.333, 0.666, 1.0])
+        descritized_target = torch.bucketize(avg_target, channel_bins, right=True)
+        descritized_target = F.one_hot(descritized_target, self.output_channel_bits)
```
Comment (lucidrains) [inline vit_pytorch/mpp_pytorch.py:45]: nit, spelled discretized

File: vit_pytorch/vit_pytorch.py [hunk 9]
```diff
@@ -126,4 +126,4 @@ def forward(self, img, mask = None):
         x = x.mean(dim = 1) if self.pool == 'mean' else x[:, 0]
 
         x = self.to_latent(x)
-        return self.mlp_head(x)
+        return self.mlp_head(x)
```
Comment (lucidrains) [inline vit_pytorch/vit_pytorch.py:129]: oops, best to have a new line as the last character in a file

**Comment by zankner** [comment]
> @zankner looking better! i think remaining is really discretizing the RGB values to 512, and various cleanup and readme :)
> 
> hmm, I like my style the best :D perk of owning your own repos in a non-corporate environment ;)

@lucidrains I might be wrong but I think the discretizing RGB values is finished. I take the mean RGB over the entire patch, and then discretize each channel to 3 bits. So in the end there are 8 possible values per channel, meaning 512 possible total colors.

**Comment by zankner** [comment]
@lucidrains Hey, just wanted to check if there is anything you still need me to change?

**Comment by lucidrains** [comment]
@zankner Hi Zach! Sorry for the delay in response, had been busy

Could you point me in the paper where they show they are doing the binary cross entropy over the 8 bits of each token? I still think what they are doing is discretizing to 512 values for each token, and doing a direct cross entropy on that, but I could be wrong

**Comment by zankner** [comment]
@lucidrains by token are you referring to the whole patch of the image ?

**Comment by lucidrains** [comment]
@zankner yup, each token, representing a patch

**Comment by zankner** [comment]
@lucidrains I was mostly looking at section B.1.2 in the appendix. When they said "predicting only the mean,
3bit color (i.e., 1 prediction of 512 colors)" I took this to mean that for each pixel in the patch, you would discretize each channel to 3-bits. Across 3 channels this would lead to 512 possible colors per pixel. Then you would take the mean of the image patch, and that would serve as the target. The transformer would then try to predict this mean of the discretized values.

Still not completely convinced of this so I would be interested to hear your interpretation.

**Comment by lucidrains** [comment]
I think you can use something like https://stackoverflow.com/questions/55918468/convert-integer-to-pytorch-tensor-of-binary-bits to convert from binary to decimal, then treat that as a hard label to predict (512 values)

**Comment by zankner** [comment]
@lucidrains Oh ok yeah that makes a lot of sense. With respect to the mean, I should take the mean before I convert to binary right?

**Comment by lucidrains** [comment]
@zankner yup!

**Comment by zankner** [comment]
@lucidrains added the converting bin to decimal. If it looks good I can add a section to the read me

**Comment by lucidrains** [comment]
@zankner this looks good now! if you write up a section in the readme, happy to merge this in 💯 💯

**Comment by zankner** [comment]
@lucidrains Thanks a lot for all your help on this! If there is any other project you are working on that you would want my help with please let me know, would love to work on it!
